### ISSUE TITLE
Undocumented behaviour with regards to attributeChangedCallback and DOM access

### ISSUE THREAD

**Issue by ruphin** [WICG/webcomponents#767]
According to the specification, the limitation of not allowing access to children or attributes of a Custom Element [only applies to the constructor](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-conformance). 

At the moment, `attributeChangedCallback` may be called immediately after the `constructor` during the lifecycle of a Custom Element, as demonstrated in [this example](https://codepen.io/ruphin/pen/dgVgJq?editors=1010). In this case, the children and attributes of the element may also not be available. 

The current implementations in Chrome, Safari, and Firefox all show the same behaviours with regards to this issue, so the behaviour seems to be specified, but I think it would be good to have it explicitly documented. We stumbled upon this issue when developing a DOM rendering framework for use with Web Components, and it would be nice to have explicit clarification from the spec.

Perhaps we can attach a note to step 4 in [the upgrade path](https://html.spec.whatwg.org/multipage/custom-elements.html#upgrades) that documents this behaviour. I can make a PR for this if requested.

**Comment by domenic** [comment]
The attributes should be available. I don't see your gist demonstrating otherwise.

Otherwise, the fact that children are not available makes sense; it's the attribute changed callback, not the children-are-available callback. I'm not sure why we'd need to document all the unrelated things that might not have happened yet.

**Comment by ruphin** [comment]
I updated the test to show that indeed all attributes are available. I first thought their availability would depend on the order in which the callback is called, but it looks like all attributes are available.

I agree that the fact children are not available makes sense, and I don't think the behaviour is wrong. I think it's more an issue with the clarity of the spec.

The issue we have is that we are using `document.importNode()` on the content of a Template Element, which causes custom elements in the newly created DocumentFragment to upgrade. Our code depends on the DOM not changing shape during this process, which was (or so we thought) a reasonable assumption since manipulating children is not allowed in the `constructor`. However, since there is nothing indicating that there are cases where it is not allowed in the `attributeChangedCallback`, it is possible that developers accidentally manipulate the children during this callback, which causes our code to break. More clarity from the spec may help alleviate this issue.

As a developer, the current spec reads like "during the constructor, children are not available, but everywhere else, they are", since it is [explicitly documented as a special case for the constructor](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-conformance), but not mentioned anywhere else. I think it would be beneficial to have some sort of note explaining that children may not be yet be attached at other points in the lifecycle.

**Comment by ruphin** [comment]
Perhaps I should rephrase. Essentially what I'm asking is if the following statements are correct:

The spec explicitly does not allow manipulating childNodes in the `constructor`. For example, if you use `this.appendChild()` in the constructor of a Custom Element, the browser will throw some form of error.

However, the spec allows manipulating the childNodes of an element in the `attributeChangedCallback`, _even when the children from the main document have not been attached yet_.

If a Custom Element adds children during one of these lifecycle callbacks, any children from the main document are appended _after_ the children inserted by the Custom Element.

In other words, the following example is valid according to the spec and behaves as expected: https://codepen.io/ruphin/pen/PyJXOe?editors=1000

**Comment by domenic** [comment]
That sounds right. There's nothing in the spec that governs what you should or should not do during attributeChangedCallback.

It sounds as if you think there's some implication otherwise. But there's no connection between children and attributeChangedCallback, just like there's no connection between (e.g.) fetches and attributeChangedCallback. Sometimes children will be there; sometimes they won't. It depends on what else is going on (e.g. upgrades, createElement vs. constructor vs. HTML parser, etc.). Just like sometimes there will be ongoing fetches, and sometimes there won't be.

If there's something in the spec implying such a connection, that'd be good to straighten out. But note that saying things about the constructor does not have any implications, positive or negative, for attributeChangedCallback, or for fetch callbacks, or for keydown listeners, or any other such things.

**Comment by ruphin** [comment]
Ok, thank you for the clarification.
